### ISSUE TITLE
Emit events no matter if method has been overwritten in test class or not

### ISSUE THREAD

**Issue by ogmueller** [sebastianbergmann/phpunit#5573]
Let's say, I want to write an extension, that wants to measure the time between beforeFirstTest (`setUpBeforeClass`) and afterLastTest (`tearDownAfterClass`) to have a duration comparison between test classes. In that case my expectation is, that I can just subscribe to the appropriate events, store start and end time, and calculate the duration of it. The problem is, that if a test class does NOT overwrite any of the methods (`setUpBeforeClass` or `tearDownAfterClass`), the events are not fired. In which case, the extension does not work and I can't measure the time. As an extension provider I cannot rely on people overwriting the methods for the sake of measurement and putting some "none-sense" in the method so phpunit is not disregarding it (e.g. just calling `parent::tearDownAfterClass()`).

Or another extension example: If I want to collect or output a summary of each test class using the afterLastTest event (`tearDownAfterClass`), I would have the same problem. My summary would not get printed, if the developer hasn't added a non-empty `tearDownAfterClass` method in the test class.

The examples might not be the best, but should showcase the general issue. If events are only emitted if someone is overwriting a certain method, many extension developers are a bit lost here.
As an extension developer I need to be sure, that some of the events get triggered within the "test cycle" no matter what. I don't care if `tearDownAfterClass` "does" something, I just want to do/print/measure/trigger/... something after a test class is done. 
Or at least a subscription/event should give me the option to enforce the emission.

(i believe, the non emitting behaviour was introduced in https://github.com/sebastianbergmann/phpunit/commit/d483e4a72d86de3262d50e11f1b8dff7afffafa7)

**Comment by sebastianbergmann** [comment]
An event that is emitted indicates that something has happened. We will not emit events that have not happened.

I am not sure that I understand your use case correctly, but I suspect that you are looking at the wrong events. 

If you are interested in the time spent in a test case class such as `MyTest` (shown below) then you should subscribe to the "Test Suite Started" and "Test Suite Ended" events. You would be looking for the time between

```
[00:00:00.074918522 / 00:00:00.000072340] [4194304 bytes] Test Suite Started (MyTest, 1 test)
```

and

```
[00:00:02.100173042 / 00:00:00.000112987] [4194304 bytes] Test Suite Finished (MyTest, 1 test)
```

#### `MyTest.php`
```php
<?php declare(strict_types=1);
use PHPUnit\Framework\TestCase;

final class MyTest extends TestCase
{
    public static function setUpBeforeClass(): void
    {
        sleep(1);
    }

    public static function tearDownAfterClass(): void
    {
        sleep(1);
    }

    public function testOne(): void
    {
        $this->assertTrue(true);
    }
}
```
```
$ phpunit --no-output --log-events-verbose-text php://stdout MyTest.php
[00:00:00.000254931 / 00:00:00.000012310] [4194304 bytes] PHPUnit Started (PHPUnit 10.4.2-36-g538b2376b6 using PHP 8.2.13 (cli) on Linux)
[00:00:00.047262045 / 00:00:00.047007114] [4194304 bytes] Test Runner Configured
[00:00:00.066217319 / 00:00:00.018955274] [4194304 bytes] Test Suite Loaded (1 test)
[00:00:00.073740916 / 00:00:00.007523597] [4194304 bytes] Event Facade Sealed
[00:00:00.074077112 / 00:00:00.000336196] [4194304 bytes] Test Runner Started
[00:00:00.074378212 / 00:00:00.000301100] [4194304 bytes] Test Suite Sorted
[00:00:00.074846182 / 00:00:00.000467970] [4194304 bytes] Test Runner Execution Started (1 test)
[00:00:00.074918522 / 00:00:00.000072340] [4194304 bytes] Test Suite Started (MyTest, 1 test)
[00:00:00.075864157 / 00:00:00.000945635] [4194304 bytes] Before First Test Method Called (MyTest::setUpBeforeClass)
[00:00:01.076462690 / 00:00:01.000598533] [4194304 bytes] Before First Test Method Finished:
                                                          - MyTest::setUpBeforeClass
[00:00:01.083987772 / 00:00:00.007525082] [4194304 bytes] Test Preparation Started (MyTest::testOne)
[00:00:01.084270410 / 00:00:00.000282638] [4194304 bytes] Test Prepared (MyTest::testOne)
[00:00:01.086981738 / 00:00:00.002711328] [4194304 bytes] Assertion Succeeded (Constraint: is true, Value: true)
[00:00:01.089833555 / 00:00:00.002851817] [4194304 bytes] Test Passed (MyTest::testOne)
[00:00:01.099412748 / 00:00:00.009579193] [4194304 bytes] Test Finished (MyTest::testOne)
[00:00:02.099821820 / 00:00:01.000409072] [4194304 bytes] After Last Test Method Called (MyTest::tearDownAfterClass)
[00:00:02.100060055 / 00:00:00.000238235] [4194304 bytes] After Last Test Method Finished:
                                                          - MyTest::tearDownAfterClass
[00:00:02.100173042 / 00:00:00.000112987] [4194304 bytes] Test Suite Finished (MyTest, 1 test)
[00:00:02.101494647 / 00:00:00.001321605] [4194304 bytes] Test Runner Execution Finished
[00:00:02.101642564 / 00:00:00.000147917] [4194304 bytes] Test Runner Finished
[00:00:02.106636042 / 00:00:00.004993478] [4194304 bytes] PHPUnit Finished (Shell Exit Code: 0)
```

In the example shown above, the test runner is invoked with the path to a test case class' source file as an argument. This is why there is only one "Test Suite Started" and "Test Suite Ended" event. There will be a pair "Test Suite Started" and "Test Suite Ended" events for each test case class.

If you are running a test suite based on an XML configuration file, [for instance this one](https://github.com/sebastianbergmann/raytracer), then you will see something like this:

```
[00:00:00.108593805 / 00:00:00.000639297] [4194304 bytes] Test Runner Execution Started (172 tests)
[00:00:00.108838844 / 00:00:00.000245039] [4194304 bytes] Test Suite Started (/usr/local/src/raytracer/phpunit.xml, 172 tests)
[00:00:00.109097024 / 00:00:00.000258180] [4194304 bytes] Test Suite Started (unit, 172 tests)
[00:00:00.109324404 / 00:00:00.000227380] [4194304 bytes] Test Suite Started (SebastianBergmann\Raytracer\CameraTest, 7 tests)
[00:00:00.112166019 / 00:00:00.002841615] [4194304 bytes] Test Preparation Started (SebastianBergmann\Raytracer\CameraTest::test_the_pixel_size_for_a_horizontal_canvas)
[00:00:00.112633397 / 00:00:00.000467378] [4194304 bytes] Test Prepared (SebastianBergmann\Raytracer\CameraTest::test_the_pixel_size_for_a_horizontal_canvas)
[00:00:00.116307021 / 00:00:00.003673624] [4194304 bytes] Assertion Succeeded (Constraint: is equal to 0.01 with delta <0.000000>, Value: 0.009999999999999998)
[00:00:00.116495293 / 00:00:00.000188272] [4194304 bytes] Test Passed (SebastianBergmann\Raytracer\CameraTest::test_the_pixel_size_for_a_horizontal_canvas)
[00:00:00.117216068 / 00:00:00.000720775] [4194304 bytes] Test Finished (SebastianBergmann\Raytracer\CameraTest::test_the_pixel_size_for_a_horizontal_canvas)
```

In the output shown above, you can see a "Test Suite Started" event for the top-level test suite configured using `phpunit.xml`, one "Test Suite Started" event for the test suite configured using an `<testsuite>` element in that XML configuration file, and one "Test Suite Started" event for the `CameraTest` test case class. TL;DR: You need to take this nesting into account.

I hope this helps.

**Comment by ogmueller** [comment]
Yes, that explanation helped. Thanks.

I was able to solve my issue with my extension by using the `PHPUnit\Event\TestSuite\FinishedSubscriber` and checking the event like this `$event->testSuite()->isForTestClass()`. So, I was able to limit the test suite selection to a single test class. It was not clear to me, that a test suite is defined in those 3 ways you described.

I would still argue that it makes sense to be able to get into the test cycle at any point of “time” for the purpose of writing extensions. Other event systems do that, too, though I also do see your point. 

I would have strongly expected it to fire to a `AfterLastTestMethodCalledSubscriber` which is related to the `tearDownAfterClass` method, no matter if I was overwriting it or not. And I would have expected that because the method itself exists in all testcases. All testcases inherit from `PHPUnit\Framework\TestCase`, which has a method called `tearDownAfterClass`. And if I put an empty ` tearDownAfterClass` method into the test class, it also fired. Even though it doesn’t do anything. So, this behavior is inconsistent.

Though, I personally don’t have a good example for this case myself, I could still imagine that extension developers would like to “do or print” something after the last test within a test class but before something like ` tearDownAfterClass` no matter if a class has overwritten it. That is just not possible right now.

I guess, we just disagree on that one. Maybe someone else need to come up with a practical use case.

Thanks again for helping me with my extension issue.
